awswrangler/lakeformation/_read.py

```diff
@@ -0,0 +1,348 @@
+
+from awswrangler import _data_types, _utils, catalog, exceptions
+from awswrangler._config import apply_configs
+from awswrangler.catalog._utils import _catalog_id
+from awswrangler.lakeformation._utils import abort_transaction, begin_transaction, wait_query
+
+_logger: logging.Logger = logging.getLogger(__name__)
+
+
+def _execute_query(
+    query_id: str,
+    token_work_unit: Tuple[str, int],
+    categories: Optional[List[str]],
+    safe: bool,
+    map_types: bool,
+    use_threads: bool,
+    boto3_session: boto3.Session,
+) -> pd.DataFrame:
+    client_lakeformation: boto3.client = _utils.client(service_name="lakeformation", session=boto3_session)
+    token, work_unit = token_work_unit
+    messages: NativeFile = client_lakeformation.execute(QueryId=query_id, Token=token, WorkUnitId=work_unit)["Messages"]
+    table: Table = RecordBatchStreamReader(messages.read()).read_all()
+    args: Dict[str, Any] = {}
+    if table.num_rows > 0:
+        args = {
+            "use_threads": use_threads,
+            "split_blocks": True,
+            "self_destruct": True,
+            "integer_object_nulls": False,
+            "date_as_object": True,
+            "ignore_metadata": True,
+            "strings_to_categorical": False,
+            "categories": categories,
+            "safe": safe,
+            "types_mapper": _data_types.pyarrow2pandas_extension if map_types else None,
+        }
+    df: pd.DataFrame = _utils.ensure_df_is_mutable(df=table.to_pandas(**args))
+    return df
+
+
+def _resolve_sql_query(
+    query_id: str,
+    categories: Optional[List[str]],
+    safe: bool,
+    map_types: bool,
+    use_threads: bool,
+    boto3_session: boto3.Session,
+) -> Union[pd.DataFrame, Iterator[pd.DataFrame]]:
+    client_lakeformation: boto3.client = _utils.client(service_name="lakeformation", session=boto3_session)
+
+    wait_query(query_id=query_id, boto3_session=boto3_session)
+
+    # The LF Query Engine distributes the load across workers
+    # Retrieve the tokens and their associated work units until NextToken is ''
+    # One Token can span multiple work units
+    # PageSize determines the size of the "Units" array in each call
+    scan_kwargs: Dict[str, Union[str, int]] = {"QueryId": query_id, "PageSize": 10}
+    next_token: str = "init_token"  # Dummy token
+    token_work_units: List[Tuple[str, int]] = []
+    while next_token:
+        response = client_lakeformation.get_work_units(**scan_kwargs)
+        token_work_units.extend(  # [(Token0, WorkUnitId0), (Token0, WorkUnitId1), (Token1, WorkUnitId2) ... ]
+            [
+                (unit["Token"], unit_id)
+                for unit in response["Units"]
+                for unit_id in range(unit["WorkUnitIdMin"], unit["WorkUnitIdMax"] + 1)  # Max is inclusive
+            ]
+        )
+        next_token = response.get("NextToken", None)
+        scan_kwargs["NextToken"] = next_token
+
+    dfs: List[pd.DataFrame] = list()
+    if use_threads is False:
```

Comment:
Just a thought: we seem to do something similar in many places, some potential for refactoring here.